Feature/ab#57 family registration wizard

Joren-Thijs opened the pull request:
[AB#57](https://dev.azure.com/berendwouters/0d4dcc20-a52e-4b6e-90d1-33c5d5079f23/_workitems/edit/57) Family registration wizard

- Reorganized and separated Dashboard components
- Added Admin Menu on dashboard
- Added add Family Wizard
- Added new Family Wizard Logo's
- Added detail user cards to Shared Module
- Moved user card to Shared Module
- Implemented BEM naming convention for CSS
- Modified Detail Forms to optionally hide the user linking form
- Modified Agegroup chips to accept single agegroup input as well
- Improved CSS
- Improved Dashboard Layout
- Removed Debugger statement
- Formatted code

![afbeelding](https://user-images.githubusercontent.com/41837971/69100509-a1bd5e80-0a5d-11ea-8d6c-e0921723d390.png)

Joren-Thijs committed:
feat(family-registration-wizard) reorganised
dashboard modules

AB#57

Joren-Thijs committed:
feat(family-registration-wizard) added
authentication service.

AB#57

Joren-Thijs committed:
Merge branch 'develop' into feature/AB#57-family-registration-wizard

Joren-Thijs committed:
feat(wizard) separated dashboard components

AB#57

Joren-Thijs committed:
feat(wizard) added wizard markup and styling
added wizard template content and data classes
added wizard data source to handle API calls

AB#57

Joren-Thijs committed:
feat(wizard) implemented datasource
expanded markup template
added forms
added Material outlined icons
expanded AddFamilyWizardStep Class
implemented re-routing
implemented error handling

AB#57

Joren-Thijs committed:
feat(wizard) removed datasource
add user implemented logic
implemented linking user logic
implemented Add-button
styled add-button
added background images

AB#57

Joren-Thijs committed:
feat(wizard) refacored HTML

AB#57

Joren-Thijs committed:
feat(wizard) added detailed user card component
removed uneccesary method from legalguardian service

AB#57

Joren-Thijs committed:
feat(wizard) implemented detailed-user-card
added imports and declarations for detailed user card
added Input and outputs to user card

implemented CSS grid layout for user card

AB#57

Joren-Thijs committed:
feat(wizard) added user card layout styling

AB#57

Joren-Thijs committed:
feat(wizard) finalysed user card grid layout

AB#57

Joren-Thijs committed:
feat(wizard) finished detailed user card styles

added color input to detailed user card

AB#57

Joren-Thijs committed:
feat(wizard) detailed user card small fixes

AB#57

Joren-Thijs committed:
feat(wizard) expanded agegroup chips
agegroup chips now accept single agegroup input

AB#57

Joren-Thijs committed:
feat(wizard) prevented text overflow on user cards

AB#57

Joren-Thijs committed:
feat(wizard) added content divider

AB#57

Joren-Thijs committed:
feat(wizard) finalized add family wizard
added NgFor for user cards
added background images
replaced add button icon images
added CSS table of contents

AB#57

Joren-Thijs committed:
feat(wizard) formatted code

AB#57

Joren-Thijs committed:
feat(wizard) modified detail forms

AB#57

Joren-Thijs committed:
Merge branch 'develop' into feature/AB#57-family-registration-wizard

AB#57

Joren-Thijs committed:
feat(wizard) merge from develop fixes

AB#57

Joren-Thijs committed:
feat(wizard) finalized family wizard
added color enums to project
made buttons on detail user card disapear if event is not used
formatted code

AB#57

Joren-Thijs committed:
feat(wizard) code formatting

AB#57

Joren-Thijs committed:
feat(wizard) removed delete events from user cards
in wizard
formatted code

AB#57

Joren-Thijs committed:
feat(wizard) renamed admin menu button

AB#57

Joren-Thijs committed:
feat(wizard) CSS: added admin menu padding

AB#57

vyruz1986 commented on Singer.API/ClientApp/src/app/modules/core/core.module.ts:
```diff
@@ -42,7 +42,8 @@ export const MY_FORMATS = {
       SingleRegistrationComponent,
-      DailybasisRegistrationsComponent
+      DailybasisRegistrationsComponent,
+      DetailedUserCardComponent,
```
This component should belong to the `SharedModule` instead of the `CoreModule`.
Same is true for the `UserCardComponent`, but when that was created we didn't have a `SharedModule` yet

Joren-Thijs replied:
The components have been moved to the Shared Module.

vyruz1986 reviewed (changes requested):
Only reviewed 24/62 files, but need to go to work :) Will review the other files later today

Joren-Thijs committed:
feat(wizard) fixed user card color bindings

added warn color variants to sint-gerardus-theme.scss

Added own material icons to project

AB#57

Joren-Thijs committed:
feat(wizard) moved user cards to shared module

small CSS fix for detail user card

AB#57

Joren-Thijs committed:
feat(wizard) bugfix:
edited user in wizard was duplicated

AB#57

Joren-Thijs committed:
feat(wizard) improved wizard add button,
now uses only one image

AB#57

Joren-Thijs committed:
feat(wizard) updated dashboard styling

AB#57

vyruz1986 commented on Singer.API/ClientApp/src/app/modules/core/services/legal-guardians-api/legalguardians.service.ts:
```diff
@@ -6,6 +6,7 @@ import {
    UpdateLegalGuardianDTO,
    CreateLegalGuardianDTO,
    LegalGuardian,
+   LegalGuardianDTO,
```
This is an unnecessary include, please remove

Joren-Thijs replied:
Removed.

vyruz1986 commented on Singer.API/ClientApp/src/app/modules/core/models/singer-colors.ts:
```diff
@@ -0,0 +1,154 @@
+/*
```
@Joren-Thijs : What's the point of having TS class for the singer colors? Why can't we use the SCSS classes we had before?

Joren-Thijs replied:
Because I can't acces SCSS variables from inside TS directly without using hacks, the point of this class is to have the colors available in TS too. In our current use case the detailed user cards have an HTML color atribute that can be used to set the background color of the Little initials circle at the top. These TS variables and functions help us do this in a handy way providing customization and randomizing.

Joren-Thijs replied:
This is the alternative: https://stackoverflow.com/questions/40418804/access-sass-values-colors-from-variables-scss-in-typescript-angular2-ionic2

It's not that bad but i preffered the namespace and the use of enums because thay make picking color values trough a function based on numbers like an index or a random number easy. If neccessary this can be changed too.

I do see the point in only defining colors in one place, maybe over time we can implement a hybrid solution?

vyruz1986 replied:
I don't know of a better way to accomplish this right now, so perhaps we can leave it as is for now. @BerendWouters maybe you have another insight?

Joren-Thijs replied:
I have an idea💡

vyruz1986 replied:
Nice 👍

vyruz1986 commented on Singer.API/ClientApp/src/app/modules/dashboard/components/shared-components/event-list/event-list.component.ts:
```diff
@@ -27,11 +27,19 @@ export class HomeComponent implements OnInit {
          .subscribe(res => {
             this.availableLocations = res.items as SingerEventLocation[];
          });
-      this._authService.isAuthenticated$.subscribe(res => {
```
Why was the authentication check removed?!

Joren-Thijs replied:
Because there is already an authentication check happening in the dashboard component where this component is being used. Double authentication seemed uneccesary.

vyruz1986 replied:
clear, thanks

vyruz1986 commented on Singer.API/ClientApp/src/app/modules/shared/components/detailed-user-card/detailed-user-card.component.html:
```diff
@@ -0,0 +1,60 @@
+      <div class="card-content">
+
+         <div class="head">
+            <div class="head__initials" [ngStyle]="{'background': color}">
```
Why do we have class names with double underscores?

Joren-Thijs replied:
Because I Implemented the BEM (Block Element Modifier) Naming convention for our CSS to make it more readable and more organized.

Joren-Thijs replied:
example: 
class="head__initials color--white"
class="Block__Element Modifier"

vyruz1986 replied:
OK, I think it's a good idea in itself, but I don't like that we now have 2 naming conventions across this project in our CSS files. Please create a task to clean this up in the next sprint

BerendWouters replied:
Right, this answers my question too.

vyruz1986 commented on Singer.API/ClientApp/src/app/modules/shared/components/detailed-user-card/detailed-user-card.component.ts:
```diff
@@ -0,0 +1,25 @@
+})
+export class DetailedUserCardComponent implements OnInit {
+   @Input() user: any = {};
+   @Input() color: string = '#2a73d4'; //Default color = primary theme color
```
Hardcoded color value? Why not in SCSS?

Joren-Thijs replied:
Again i do not have acces to SCSS variables in a direct and clean way without using hacks, this color value is just a default for when the user doesn't specify one. Altough you are correct that in this perticular instance it is unnesecary and should be replaced by a CSS Default. And that the Color variable should only be applied with NgStyle when used. I will change this tommorow as well.

vyruz1986 reviewed (changes requested):
Encountered following issues during testing:
* Admin menu is currently on dashboard page, is that the intended location? It doesn't seem right to me...
* Dashboard layout has been changed, event cards have changed, they are now in a single column and very wide... This should not have been changed in this branch
* Why have several components been moved into the dashboard module under a shared-components folder? If they should be shared globally, they should be in the Shared module, and if not, they don't need to be in a shared-components subfolder

Please also address or discuss the inline comments given on the file changes

Joren-Thijs commented:
> 
> 
> Encountered following issues during testing:
> 
>     * Admin menu is currently on dashboard page, is that the intended location? It doesn't seem right to me...
> 
>     * Dashboard layout has been changed, event cards have changed, they are now in a single column and very wide... This should not have been changed in this branch
> 
>     * Why have several components been moved into the dashboard module under a shared-components folder? If they should be shared globally, they should be in the Shared module, and if not, they don't need to be in a shared-components subfolder
> 
> 
> Please also address or discuss the inline comments given on the file changes

Dashboard Page was the chosen place for the admin menu, i asked and Berend approved. If desired this can be moved.

Dashboard max-width has been reduced to 1200px, this is considerd part of the admin menu changes. Event cards appear very wide but reset to small on window resize, this bug is unrelated to the change in total width and is caused by the recent PR [AB#112](https://dev.azure.com/berendwouters/0d4dcc20-a52e-4b6e-90d1-33c5d5079f23/_workitems/edit/112) (Expected to be updated tommorow).

Dashboard components where originally in a single folder but since the introduction of the admin menu we have components that are only to be viewed by admin users, potentially only by legalguardian users, and components that are shared between all  users like the event list.
If desired this can be changed.

Joren-Thijs committed:
feat(wizard) removed unnused import

AB#57

Joren-Thijs committed:
Merge branch 'develop' into feature/AB#57-family-registration-wizard

vyruz1986 commented:
> > * Why have several components been moved into the dashboard module under a shared-components folder? If they should be shared globally, they should be in the Shared module, and if not, they don't need to be in a shared-components subfolder
> > 
> Dashboard components where originally in a single folder but since the introduction of the admin menu we have components that are only to be viewed by admin users, potentially only by legalguardian users, and components that are shared between all users like the event list.
> If desired this can be changed.

You're missing the point :) In angular the structure should be one of these 2 options:
* Components are 'shared' by default within the module they are defined in, so no need to create another 'shared-components' folder for components which are only used in the same module (in this case the DashboardModule)
* Components which need to be reused outside of the module they are defined in, should be moved to the `SharedModule`, so they can reused in other modules.

I see what you're trying to do, organising the components into admin, shared (and potentially legal-guardians in the future), however that makes no sense from an angular architecture perspective. Currently the Dashboard is one module for all user types, if there is a need to separate the module, it should be split at the module level (thus creating a `AdminDashboardModule`, `LegalGuadianDashboardModule`, and putting shared components in `SharedModule`)

For now, I suggest putting all components for the `DashboardModule` in a `components` subfolder

Joren-Thijs committed:
feat(wizard) moved dashboard components
back into general components folder

AB#57

Joren-Thijs committed:
Merge branch 'develop' into feature/AB#57-family-registration-wizard

AB#57

Joren-Thijs committed:
feat(wizard) merge from develop

AB#57

Joren-Thijs commented:
> 
> 
> > > * Why have several components been moved into the dashboard module under a shared-components folder? If they should be shared globally, they should be in the Shared module, and if not, they don't need to be in a shared-components subfolder
> > 
> > 
> > Dashboard components where originally in a single folder but since the introduction of the admin menu we have components that are only to be viewed by admin users, potentially only by legalguardian users, and components that are shared between all users like the event list.
> > If desired this can be changed.
> 
> You're missing the point :) In angular the structure should be one of these 2 options:
> 
>     * Components are 'shared' by default within the module they are defined in, so no need to create another 'shared-components' folder for components which are only used in the same module (in this case the DashboardModule)
> 
>     * Components which need to be reused outside of the module they are defined in, should be moved to the `SharedModule`, so they can reused in other modules.
> 
> 
> I see what you're trying to do, organising the components into admin, shared (and potentially legal-guardians in the future), however that makes no sense from an angular architecture perspective. Currently the Dashboard is one module for all user types, if there is a need to separate the module, it should be split at the module level (thus creating a `AdminDashboardModule`, `LegalGuadianDashboardModule`, and putting shared components in `SharedModule`)
> 
> For now, I suggest putting all components for the `DashboardModule` in a `components` subfolder

Changes have been made.

BerendWouters commented on Singer.API/ClientApp/src/app/modules/admin/components/add-family-wizard/add-family-wizard.component.html:
```diff
@@ -0,0 +1,115 @@
+                  <ng-container *ngIf="i == 0">
+                     <div class="body__img-container">
+                        <img src="../../../../../assets/svg/FamilyWizardStartLogo.svg" alt="Achtergrond Afbeelding"
+                           title="Achtergrond Afbeelding" class="img-container__img">
```
Why the `__`?

BerendWouters commented on Singer.API/ClientApp/src/app/modules/admin/components/add-family-wizard/add-family-wizard.component.ts:
```diff
@@ -0,0 +1,312 @@
+      {
+         stepLabel: 'Start',
+         topText:
+            'Welkom bij de Familie toevoegen wizard, hier kan je gemakkelijk niewe Voogden en Zorggebruikers gelijktijdig inschrijven.',
```
`Voogden` and `Zorggebruikers` don't need capitals

BerendWouters commented on Singer.API/ClientApp/src/app/modules/admin/components/add-family-wizard/add-family-wizard.component.ts:
```diff
@@ -0,0 +1,312 @@
+   linkUsers(): boolean {
+      //collect all the careUser id's
+      var careUsersToAdd: string[] = [];
+      this.careUsers.forEach(careUser => {
```
Why don't you just do the following?
`const careUsersToAdd = this.careUsers.map(careUser => careUser.id);`
?

Joren-Thijs replied:
I didn't know that was an option😅 It's fixed now.

BerendWouters commented on Singer.API/ClientApp/src/app/modules/dashboard/components/event-list/event-list.component.ts:
```diff
@@ -1,43 +1,50 @@
    }
 
    onResize(event) {
       this.breakpoint = event.target.innerWidth <= 500 ? 1 : 3;
```
You could be able to get the current window in TS too OnInit

Joren-Thijs replied:
Yes this was causing that giant cards until resized bug. Fixed it. Thx!

BerendWouters commented on Singer.API/ClientApp/src/app/modules/dashboard/components/home/home.component.ts:
```diff
@@ -1,42 +1,24 @@
    templateUrl: './home.component.html',
    styleUrls: ['./home.component.css'],
 })
 export class HomeComponent implements OnInit {
```
If the ngOnInit() is empty, you could remove the OnInit implementation too.

Joren-Thijs replied:
Done.

BerendWouters reviewed (changes requested):
Some general remarks that needs processing, but I see some nice things.
One additional remark: why the specific Material Icons in the assets?

Joren-Thijs committed:
feat(wizard) removed singer-colors.ts

added additional accents to sint-gerardus-theme.scss
fixed comments
formatted code
added new random background color selection method to user cards

AB#57

Joren-Thijs committed:
feat(wizard) Removed Material icons
 that are currently unused

AB#57

Joren-Thijs committed:
feat(wizard) fixed typo's

AB#57

Joren-Thijs committed:
feat(wizard) fixed event cards layout bug

AB#57

Joren-Thijs committed:
feat(wizard) optemized code

AB#57

Joren-Thijs committed:
feat(wizard) Removed unused ngOninit
Implementations

AB#57

Joren-Thijs commented:
> 
> 
> Some general remarks that needs processing, but I see some nice things.
> One additional remark: why the specific Material Icons in the assets?

Because using the mat-icon element from angular material does not provide a responsive icon. The icons are stuck at a single size. The only way to make them bigger is to apply a CSS transform to them wich breaks their place inside the document flow and does not keep them centered, making them a layout nightmare. The plan is to replace them by responsive images wherever possible.

The reason we had so many was i experimented with them in the detail user card, however there was an issue with them not loading in on time (1.5 sec delay) even tough they were already cached by the browser. And since i was not able to solve this quickly i reverted them back to mat-icon. However i plan on another try in the future.

For now most of them are uneccesarry so they are removed from this PR. Except for the add box wich is currently used to form the wizards giant add button.

Joren-Thijs committed:
Merge branch 'develop' into feature/AB#57-family-registration-wizard

AB#57
